flink-runtime/src/main/java/org/apache/flink/runtime/state/v2/InternalReducingState.java

```diff
@@ -58,6 +72,108 @@ public V get() {
+            return StateFutureUtils.completedVoidFuture();
+        }
+        List<StateFuture<V>> futures = new ArrayList<>();
+        for (N source : sources) {
+            if (source != null) {
+                setCurrentNamespace(source);
+                futures.add(handleRequest(StateRequestType.REDUCING_GET, null));
+            }
+        }
+        return StateFutureUtils.combineAll(futures)
+                .thenCompose(
+                        values -> {
+                            List<StateFuture<V>> removeFutures = new ArrayList<>();
+                            V current = null;
+                            Iterator<N> sourceIter = sources.iterator();
+                            for (V value : values) {
+                                N source = sourceIter.next();
+                                if (value != null) {
+                                    setCurrentNamespace(source);
+                                    removeFutures.add(
+                                            handleRequest(StateRequestType.REDUCING_REMOVE, null));
+                                    if (current != null) {
+                                        current = reduceFunction.reduce(current, value);
+                                    } else {
+                                        current = value;
+                                    }
+                                }
+                            }
+                            V finalCurrent = current;
+                            return StateFutureUtils.combineAll(removeFutures)
+                                    .thenApply(ignore -> finalCurrent);
+                        })
+                .thenAccept(
+                        currentValue -> {
+                            if (currentValue == null) {
+                                return;
+                            }
+                            setCurrentNamespace(target);
+                            handleRequest(StateRequestType.REDUCING_GET, null)
+                                    .thenAccept(
+                                            targetValue -> {
+                                                if (targetValue == null) {
+                                                    handleRequest(
+                                                            StateRequestType.REDUCING_ADD,
+                                                            currentValue);
+                                                } else {
+                                                    handleRequest(
+                                                            StateRequestType.REDUCING_ADD,
+                                                            reduceFunction.reduce(
+                                                                    currentValue, (V) targetValue));
+                                                }
+                                            });
+                        });
```

Comment:
Optimization Suggestion: Can we split all requests into two phase, GET(including the target's and sources' state split) and UPDATE?

Reply:
👍 done.